### PR TITLE
fix memory and file handling

### LINKED ISSUE #290: Out of memory error when downloading large dataset
When trying to install the new eBird Observational Dataset (#286) I run into an out of memory error on my laptop.

```
ethan@oryx:~/$ retriever install postgres eBirdOD
=> Installing eBird Observation Dataset
Creating database eBirdOD...
Couldn't create database (schema "ebirdod" already exists
). Trying to continue anyway.
Downloading CLOEODDATA.05192014.1...
out of memory
```

This is presumably because `download_file()` in `engine.py` is reading in the entire file and then writing it. This means that we need to read and write more manageable chunks.

### PR DESCRIPTION
**Pull request by jamesaq**
fixes #290

### CODE REVIEW

File: lib/engine.py [hunk 1]
```diff
@@ -364,18 +364,25 @@ def database_name(self, name=None):
...
-                local_file.write(file.read())
-            local_file.close()
-            file.close()
-
+                path = self.format_filename(filename)
+                urllib.urlretrieve(url, path) 
+
+    def clean_line_endings(self, tmp_path, filename):
+        """Writes new file with decoded newlines and removes old file."""
+        with open(tmp_path, 'rU') as tmp_file:
```
Comment (jamesaq) [inline lib/engine.py]: writes straight to disk

File: lib/engine.py [hunk 2]
```diff
@@ -364,18 +364,25 @@ def database_name(self, name=None):
     def download_file(self, url, filename, clean_line_endings=True):
         """Downloads a file to the raw data directory."""
         if not self.find_file(filename):
-            path = self.format_filename(filename)
             self.create_raw_data_dir()
-            print "Downloading " + filename + "..."
-            file = urllib.urlopen(url)
-            local_file = open(path, 'wb')
-            if clean_line_endings and (filename.split('.')[-1].lower() not in ["exe", "zip", "xls"]):
```
Comment (jamesaq) [inline lib/engine.py]: writes straight to disk

File: lib/engine.py [hunk 3]
```diff
@@ -364,18 +364,25 @@ def database_name(self, name=None):
...
-                local_file.write(file.read().replace("\r\n", "\n").replace("\r", "\n"))
+            print "Downloading {}...".format(filename)
+            ext = os.path.splitext(filename)[1].lower()
+            if clean_line_endings and ext not in ['exe', 'zip', 'xls']:
+                tmp_path = self.format_buffer_filename(filename)
+                urllib.urlretrieve(url, tmp_path)
```
Comment (jamesaq) [inline lib/engine.py]: writes straight to disk

File: lib/engine.py [hunk 4]
```diff
@@ -364,18 +364,25 @@ def database_name(self, name=None):
...
-                local_file.write(file.read())
-            local_file.close()
-            file.close()
-
+                path = self.format_filename(filename)
+                urllib.urlretrieve(url, path) 
+
+    def clean_line_endings(self, tmp_path, filename):
+        """Writes new file with decoded newlines and removes buffer file."""
+        with open(tmp_path, 'rU') as tmp_file:
```
Comment (jamesaq) [inline lib/engine.py]: writes straight to disk

**Comment by ethanwhite** [comment]
Thanks for this @jamesaq!

What is your thinking on cleaning line endings like this vs. switching to opening the files in universal mode as described in #291 and avoiding cleaning the line endings as all.

**Comment by jamesaq** [comment]
@ethanwhite I think using universal mode would be a better way to handle things, I added some commits

**Comment by ethanwhite** [comment]
Thanks for this @jamesaq!

It looks like we have some failing regression tests (not sure why Travis wasn't reporting them earlier, but it's fixed now):
https://travis-ci.org/weecology/retriever/builds/64184376

Take a look at them if you'd like or I can find some time later this week to see if I can figure out what's going on.

**Comment by ethanwhite** [comment]
Hi @jamesaq - just wanted to apologize for letting this languish for a month. I've been in the process of moving across the country to a new university and the software side of things has been suffering. I'm hoping to get back on top of things in the next week or two and will try to find time to run down those failing tests.

**Comment by henrykironde** [comment]
@jamesaq, thanks for your contribution. Your work was used as a reference to fix the memory issue in #502

**Pull request closed without merging**